iceoryx2-cal/src/shm_allocator/mod.rs

```diff
@@ -13,44 +13,147 @@
+    pub fn new(value: u8) -> Self {
         Self(value)
     }
 
-    pub fn value(&self) -> usize {
+    /// Returns the underlying value of the [`SegmentId`]
+    pub fn value(&self) -> u8 {
         self.0
     }
+
+    /// Returns the maximum value the [`SegmentId`] supports.
+    pub const fn max_segment_id() -> u8 {
+        u8::MAX
+    }
 }
 
-enum_gen! { ShmAllocationError
+const NUMBER_OF_BITS_IN_BYTE: usize = 8;
+
+/// An offset to a [`SharedMemory`](crate::shared_memory::SharedMemory) address. It requires the
+/// [`SharedMemory::payload_start_address()`](crate::shared_memory::SharedMemory::payload_start_address())
+/// of the corresponding [`SharedMemory`](crate::shared_memory::SharedMemory) to be converted into
+/// an actual pointer.
+///
+/// Contains the offset and the corresponding [`SegmentId`].
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub struct PointerOffset(u64);
+
+impl PointerOffset {
+    /// Creates a new [`PointerOffset`] from the given offset value with the [`SegmentId`] == 0.
+    pub fn new(offset: usize) -> PointerOffset {
+        const SEGMENT_ID: u64 = 0;
+        Self(
+            (offset as u64) << (core::mem::size_of::<SegmentId>() * NUMBER_OF_BITS_IN_BYTE)
+                | SEGMENT_ID,
+        )
+    }
+
+    /// Sets the [`SegmentId`] of the [`PointerOffset`].
+    pub fn set_segment_id(&mut self, value: SegmentId) {
+        self.0 |= value.0 as u64;
+    }
+
+    /// Returns the offset.
+    pub fn offset(&self) -> usize {
+        (self.0 >> (core::mem::size_of::<SegmentId>() * NUMBER_OF_BITS_IN_BYTE)) as usize
+    }
+
+    /// Returns the [`SegmentId`].
+    pub fn segment_id(&self) -> SegmentId {
+        SegmentId((self.0 & SegmentId::max_segment_id() as u64) as u8)
+    }
+}
+
+impl Debug for PointerOffset {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "PointerOffset {{ offset: {}, segment_id: {:?} }}",
+            self.offset(),
+            self.segment_id()
+        )
+    }
+}
+
+enum_gen! {
+/// Describes the errors that can occur when [`ShmAllocator::allocate()`] is called.
+    ShmAllocationError
   entry:
     ExceedsMaxSupportedAlignment
   mapping:
     AllocationError
 }
 
+/// Describes generically an [`AllocationStrategy`], meaning how the memory is increased when the
+/// available memory is insufficient.
+#[derive(Clone, Copy, Eq, PartialEq, Debug, Default)]
+pub enum AllocationStrategy {
+    /// Increases the memory so that it perfectly fits the new size requirements. This may lead
+    /// to a lot of reallocations but has the benefit that no byte is wasted.
+    BestFit,
+    #[default]
+    /// Increases the memory by rounding the increased memory size up to the next power of two.
+    /// Reduces reallocations a lot at the cost of increased memory usage.
+    PowerOfTwo,
+    /// The memory is not increased. This may lead to an out-of-memory error when allocating.
+    Static,
```

Comment:
Just wondering whether `Static` wouldn't be a better default? Not asking to change it, just to get an idea why you choose `PowerOfTwo`?

Reply:
You are right! This should be the default.